### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by SzymonPobiega**
* Added acceptance tests using core approach
 * Added API approval tests
 * Added input checks for the public API via `Guard` class
 * Added error handling & logging to queue reporter

As a result of going through the code I've raised discussion issues #5, #6 and #7

### CODE REVIEW

File: src/NServiceBus.Metrics.Tests/APIApprovals.ApproveNServiceBus.approved.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+﻿
```
Comment (mikeminutillo) [inline src/NServiceBus.Metrics.Tests/APIApprovals.ApproveNServiceBus.approved.txt]: What is this file for?

File: src/NServiceBus.Metrics.Tests/APIApprovals.ApproveAPI.approved.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+[assembly: System.Runtime.Versioning.TargetFrameworkAttribute(".NETFramework,Version=v4.5.2", FrameworkDisplayName=".NET Framework 4.5.2")]
+namespace NServiceBus
+{
+    public class static MetricsConfigurationExtensions
+    {
+        public static NServiceBus.MetricsOptions EnableMetrics(this NServiceBus.EndpointConfiguration endpointConfiguration) { }
+    }
+    public class MetricsOptions
+    {
+        public MetricsOptions() { }
```
Comment (mikeminutillo) [inline src/NServiceBus.Metrics.Tests/APIApprovals.ApproveAPI.approved.txt]: Should this be internalized?

File: src/NServiceBus.Metrics/MetricsFeature.cs [hunk 3]
```diff
@@ -86,5 +87,6 @@ protected override Task OnStop(IMessageSession session)
             }
         }
 
+        public static TimeSpan DefaultInterval = TimeSpan.FromSeconds(30);
```
Comment (mikeminutillo) [inline src/NServiceBus.Metrics/MetricsFeature.cs]: Why have a 3rd level default like this?

File: src/NServiceBus.Metrics/Reporters/NServiceBusMetricReport.cs [hunk 4]
```diff
@@ -26,17 +27,32 @@ public NServiceBusMetricReport(IDispatchMessages dispatcher, string destination)
...
+
+        async Task RunReportAsync(MetricsData metricsData)
+        {
             var serialized = JsonBuilderV2.BuildJson(metricsData);
             var body = Encoding.UTF8.GetBytes(serialized);
 
-            var headers = new Dictionary<string, string>();
+            var headers = new Dictionary<string, string>
+            {
+                [Headers.EnclosedMessageTypes] = "Metrics.Json.JsonMetricsContext, Metrics"
```
Comment (mikeminutillo) [inline src/NServiceBus.Metrics/Reporters/NServiceBusMetricReport.cs]: Should this be done via reflection?

File: src/NServiceBus.Metrics.AcceptanceTests/Tests/When_telemetry_queue_does_not_exist.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    using global::Metrics.Json;
+    using Logging;
+    using NServiceBus.AcceptanceTests;
+    using NServiceBus.AcceptanceTests.EndpointTemplates;
+    using NUnit.Framework;
+
+    public class When_telemetry_queue_does_not_exist : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Should_send_reports_to_configured_queue()
```
Comment (mikeminutillo) [inline src/NServiceBus.Metrics.AcceptanceTests/Tests/When_telemetry_queue_does_not_exist.cs]: Should_log_error?

File: src/NServiceBus.Metrics.AcceptanceTests/Tests/When_reporting_to_ServiceControl.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    using NServiceBus.AcceptanceTests;
+    using NServiceBus.AcceptanceTests.EndpointTemplates;
+    using NUnit.Framework;
+
+    public class When_reporting_to_ServiceControl : NServiceBusAcceptanceTest
+    {
+        static string MonitoringSpyAddress => Conventions.EndpointNamingConvention(typeof(MonitoringSpy));
+
+        [Test]
+        public async Task Should_send_reports_to_configured_queue()
```
Comment (mikeminutillo) [inline src/NServiceBus.Metrics.AcceptanceTests/Tests/When_reporting_to_ServiceControl.cs]: This test is just asserting that the message arrives. Do we need to make any sort of assertions about the data?

File: src/NServiceBus.Metrics.Tests/APIApprovals.ApproveNServiceBus.approved.txt (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Metrics.Tests/APIApprovals.ApproveNServiceBus.approved.txt]: Leftover. I'll remove.

File: src/NServiceBus.Metrics.Tests/APIApprovals.ApproveAPI.approved.txt (hunk 2 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Metrics.Tests/APIApprovals.ApproveAPI.approved.txt]: Well, `var options = endpointConfiguration.GetSettings().GetOrCreate<MetricsOptions>();` would stop working and I don't think we leak stuff by allowing people to construct an instance of this.

File: src/NServiceBus.Metrics/MetricsFeature.cs (hunk 3 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Metrics/MetricsFeature.cs]: I'll let @indualagarsamy answer. I have no strong opinions about this.

File: src/NServiceBus.Metrics/Reporters/NServiceBusMetricReport.cs (hunk 4 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Metrics/Reporters/NServiceBusMetricReport.cs]: Nope, because via reflection you would get `NServiceBus.Metrics` as assembly name.

File: src/NServiceBus.Metrics.AcceptanceTests/Tests/When_reporting_to_ServiceControl.cs (hunk 6 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Metrics.AcceptanceTests/Tests/When_reporting_to_ServiceControl.cs]: I think this should be a separate test that checks that. Here I only wanted to check if the communication works and the data can be deserialized. Contents (such as which metrics are included) should be verified elsewhere.

**Comment by mikeminutillo** [comment]
@danielmarbach is it safe to do that for the headers dictionary as well? The dispatcher shouldn't be modifying it.

**Comment by SzymonPobiega** [comment]
@danielmarbach @mikeminutillo I don't think we should do that. While it would optimize allocations, these objects are not guaranteed to be read only (probably a design mistake on my side with `TransportTransaction`).

Currently the core passes the same instance of headers to copy of the message when doing unicast pub/sub so transports should not modify the headers but in general there is nothing preventing them from doing so. 

To summarize, I am 👎 for this optimization as it can bite us in the ass down the road should anything change in the core and/or the transports.

**Comment by danielmarbach** [comment]
@mikeminutillo yes it is a raw dispatch hence there is no pipeline nothing that could modify this. The dispatcher in all cases should only read from the bags provided since it is the last step in the pipeline execution in the core. So it is safe to do this

**Comment by SzymonPobiega** [comment]
@mikeminutillo I pushed the changes and rebased it on top of develop.
